nupic/research/frameworks/vernon/mixins/gradient_metrics.py

```diff
@@ -0,0 +1,232 @@
+
+        # The targets will be collected in `self.error_loss` in a 1:1 fashion
+        # to the tensors being collected by the hooks.
+        self.gradient_metric_targets = torch.tensor([]).long()
+
+    def process_gradient_metrics_args(self, gradient_metric_args):
+
+        gradient_metrics_args = deepcopy(gradient_metric_args)
+
+        # Collect information about which modules to apply hooks to
+        include_names = gradient_metrics_args.pop("include_names", [])
+        include_modules = gradient_metrics_args.pop("include_modules", [])
+        include_patterns = gradient_metrics_args.pop("include_patterns", [])
+        filter_args = dict(
+            include_names=include_names,
+            include_modules=include_modules,
+            include_patterns=include_patterns,
+        )
+
+        # Others args
+        plot_freq = gradient_metrics_args.get("plot_freq", 1)
+        max_samples = gradient_metrics_args.get("max_samples_to_track", 100)
+        metrics = gradient_metrics_args.get("metrics", ["cosine"])
+        gradient_values = gradient_metrics_args.get("gradient_values", "real")
+
+        available_metrics_options = ["cosine", "pearson", "dot"]
+        available_gradient_values_options = ["real", "sign", "mask"]
+
+        assert isinstance(plot_freq, int)
+        assert isinstance(max_samples, int)
+        assert isinstance(metrics, list)
+        assert len(metrics) > 0
+        assert all([metric in available_metrics_options for metric in metrics])
+        assert plot_freq > 0
+        assert max_samples > 0
+        assert isinstance(gradient_values, str)
+        assert gradient_values in available_gradient_values_options
+
+        return plot_freq, filter_args, max_samples, metrics, gradient_values
+
+    def run_epoch(self):
+        """
+        This runs the epoch with the hooks in tracking mode. The resulting hidden
+        activations collected by the `TrackHiddenActivationsHook` object is plotted by
+        calling a plotting function.
+        """
+
+        # Run the epoch with tracking enabled.
+        with self.gradient_metric_hooks:
+            results = super().run_epoch()
+
+        # The epoch was iterated in `run_epoch` so epoch 0 is really epoch 1 here.
+        iteration = self.current_epoch + 1
+
+        # Calculate metrics, create visualization, and update results dict.
+        if iteration % self.gradient_metrics_plot_freq == 0:
+            gradient_stats = self.gradient_metric_hooks.get_statistics()
+            gradient_metrics_stats = self.calculate_gradient_metrics_stats(
+                gradient_stats
+            )
+            gradient_metric_heatmaps = self.plot_gradient_metric_heatmaps(
+                gradient_metrics_stats
+            )
+            for (name, _, metric, _, figure) in gradient_metric_heatmaps:
+                results.update({f"{name}/{metric}": figure})
+        return results
```

Comment:
Have you tested this with Ray? It usually doesn't respond well when you attempt to log a plot. It inevitably tries to deepcopy the plot, which can't be done since that's not implemented by matplotlib.

Reply:
One thing you can do is wrap the plot so it's in a loggable form. Say you accept an additional argument to your mixin called `prep_plots_for_wandb` (with a default False so we're wandb agnostic), then you can have logic like
```
from nupic.research.frameworks.wandb import ray_wandb

...
if self.prep_plots_for_wandb:
    figure = ray_wandb.prep_plots_for_wandb(figure)
results.update({f"{metric}/{name}": figure})
```